flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/Execution.java

```diff
@@ -566,6 +619,67 @@ public void setInitialState(@Nullable JobManagerTaskRestore taskRestore) {
 		}
 	}
 
+	@VisibleForTesting
+	CompletableFuture<Execution> registerProducedPartitions(TaskManagerLocation location) {
+		assertRunningInJobMasterMainThread();
+
+		return FutureUtils.thenApplyAsyncIfNotDone(
+			registerProducedPartitions(vertex, location, attemptId),
+			vertex.getExecutionGraph().getJobMasterMainThreadExecutor(),
+			producedPartitionsCache -> {
+				producedPartitions = producedPartitionsCache;
+				return this;
+			});
+	}
+
+	@VisibleForTesting
+	static CompletableFuture<Map<IntermediateResultPartitionID, ResultPartitionDeploymentDescriptor>> registerProducedPartitions(
+			ExecutionVertex vertex,
+			TaskManagerLocation location,
+			ExecutionAttemptID attemptId) {
+		ProducerDescriptor producerDescriptor = ProducerDescriptor.create(location, attemptId);
+
+		boolean lazyScheduling = vertex.getExecutionGraph().getScheduleMode().allowLazyDeployment();
+
+		Collection<IntermediateResultPartition> partitions = vertex.getProducedPartitions().values();
+		Collection<CompletableFuture<ResultPartitionDeploymentDescriptor>> partitionRegistrations =
+			new ArrayList<>(partitions.size());
+
+		for (IntermediateResultPartition partition : partitions) {
+			PartitionDescriptor partitionDescriptor = PartitionDescriptor.from(partition);
+			int maxParallelism = getPartitionMaxParallelism(partition);
+			partitionRegistrations.add(
+				vertex
```

Comment:
maybe define a local variable shuffle master before loop to avoid this part many layers